src/site_apportionment/site_apportionment.py

```diff
@@ -0,0 +1,219 @@
+import os
+import logging
+
+SitesApportionmentLogger = logging.getLogger(__name__)
+
+# Colunm names redefined for convenience
+ref = "reference"
+ins = "instance"
+period = "period"
+form = "formtype"
+postcode = "postcodes_harmonised"
+percent = "602"
+product = "201"
+pg_num = "pg_numeric"
+civdef = "200"
+
+# Long and short form codes
+short_code = "0006"
+long_code = "0001"
+
+
+def count_unique_codes_in_col(df: pd.DataFrame, code: str) -> pd.DataFrame:
+    """Calculates the number of unique non-empty codes in a column.
+
+    Args:
+        df (pd.DataFrame): A dataframe containing all data
+        code (str): Name of the column containing codes
+
+    Returns:
+        (pd.DataFrame): A copy of originl dataframe with an additional column
+        called the same as code with suffix "_count" countaining the number of
+        unique non-empty codes
+    """
+
+    dfa = df.copy()
+
+    # Select columns that we need
+    cols_need = [ref, period, code]
+    dfa = dfa[cols_need]
+    dfa = dfa[dfa[code].str.len() > 0]
+    dfa.drop_duplicates(inplace=True)
+    dfb = dfa.groupby([ref, period]).agg("count").reset_index()
+    dfb.rename({code: code + "_count"}, axis='columns', inplace=True)
+    df = df.merge(
+        dfb,
+        on=[ref, period],
+        how="left")
+    return df
+
+
+def weights(df):
+    """Calculates site weights based on the percents. Copies the precent value
+    from its original location (question 602) to a new column "site_percent".
+    If the percent value is Null, fills it with zero.
+
+    Adds all percents per RU reference and period and calculates
+    site_percent_total. If the total for a reference is zero, apportionment
+    cannot be done, and this reference is removed.
+    Re-noramalises the percent values by total, to compute site_weight. This
+    deals with the case when the users entered percents incorrectly, so they
+    don't sum up to 100.
+
+    Args:
+        df (pd.DataFrame): A dataframe containing all data
+
+    Returns:
+        (pd.DataFrame): A copy of original dataframe with an additional column
+        called site_weight countaining the weights of each site, between 0 and 1
+    """
+    dfc = df.copy()
+    dfc["site_percent"] = dfc[percent]
+    dfc["site_percent"].fillna(0, inplace=True)
+
+    # Set the percent for instance 0 to be 0
+    dfc["site_percent"] = dfc["site_percent"] * dfc[ins].astype("bool")
+
+    # Calculate the total percent for each reference and period
+    dfc["site_percent_total"] = (
+        dfc.groupby([ref, period])["site_percent"].transform("sum"))
+
+    # Filter out the rows where total percent is zero
+    dfc = dfc[dfc["site_percent_total"] != 0]
+
+    # Compute weights
+    dfc["site_weight"] = dfc["site_percent"] / dfc["site_percent_total"]
+
+    # Remove unnecessary
+    dfc.drop(columns=['site_percent', 'site_percent_total'], inplace=True)
+
+    return dfc
+
+
+def apportion_sites(df: pd.DataFrame)-> pd.DataFrame:
```

Comment:
The function is quite long and does a lot of different operations. It would be beneficial to break it down into smaller, more manageable functions making the code easier to read, test, and maintain.

Reply:
let's see about doing that later